### ISSUE TITLE
YML Structure Changes

### ISSUE THREAD

**Issue by Qstick** [Prowlarr/Indexers#121]
@garfield69 @ilike2burnthing

This is mostly to open discussion for yml updates given we loosely use the same definitions and I'm not keen to get to diverged from each other.

## Whats been done
- We recently added xml parsing here <https://github.com/Prowlarr/Prowlarr/commit/701c7d349ab2928ba7ce5abd342cb8870300bf19> to allow ShowRSS and come of our newznab defintions to switch to yml. This should be a pretty straight port to Jackett using the same code. 

- AllowRawSearch - Prowlarr supports using this parameter on the caps block of an indexer to specify to Sonarr or Radarr that indexer allows raw (non-cleaned) queries

## Whats in progress
- I know you guys are also looking to expand out on the json work with filters and such here <https://github.com/Jackett/Jackett/pull/12464> which will be good.

- Additionally we are planning to allow headers on login and download blocks separate from search block headers, but defaulting back to search headers. This is in PR here <https://github.com/Prowlarr/Prowlarr/pull/644>. Feel free to comment. (also a new prop to `IgnoreBlankInputs` for search block in that PR)

- IndexerFlags capabilities have been added to Prowlarr which will allow indexers to mark releases with indexer specific flags or global defined flags (freeleech, internal, etc) and pass that through in the torznab response. This allows user to better control release selection in arr. Next steps is to incorporate this in yml parsing.

## Next Steps
My next task is really to be able to add functionality to handle one item to many torrents in yml and the cardigann parser. This would allow us to starting thinking about conversion of indexers such as Gazelle based, etc.. My question to you guys before I get going is are there any thoughts or recommendation about how you would like to see this structured. 

Currently we have this
```
rows:
    selector: rss > channel > item
```

Some options are 
```
rows:
    selector: rss > channel > item

releases: 
    # In relation to row
    selector: torrents
```

or 

```
rows:
    selector: rss > channel > item
    releases: 
        # In relation to row
        selector: torrents
```

My proposal would be to just add another selector for `release` items under the row item or separate, though I've waffled about the best format for this.

at that point we could then add a `parent` prop or something similar to fields so that each field is parsed from the correct object. Some could come from the row, some from the release.

```
  fields:
    title:
      selector: title
      parent: release
```

Thoughts on this?

[AB#2397](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2397)

**Comment by bakerboy448** [comment]
tangentially related - getting a proper cardigann schema
https://github.com/Jackett/Jackett/issues/12793

**Comment by ilike2burnthing** [comment]
**xml parsing** - will give that a test for Jackett now

**AllowRawSearch** - I believe that was being discussed in https://github.com/Jackett/Jackett/issues/8246 and meant to be addressed by https://github.com/Jackett/Jackett/pull/11889, but has been held up. I don't know if @seeyabye was waiting on further feedback or it got lost.

**Cardigann v4** - will have a look at this

**IndexerFlags** - I thought this was still a pipedream for Sonarr - requested here https://github.com/Sonarr/Sonarr/issues/2782 and still in draft here https://github.com/Sonarr/Sonarr/pull/3036

**one item to many torrents** - `releases` being part of `rows`, rather than being its own block, makes the most sense to me, and the `parent` idea looks good

**cardigann schema** - zero experience here, I'd defer to you and the [Definition Format](https://github.com/Jackett/Jackett/wiki/Definition-format)

**Comment by Qstick** [comment]
> **xml parsing** - will give that a test for Jackett now
> 

Prolly need this as well <https://github.com/Prowlarr/Prowlarr/commit/470b751f4453bc575aa65f7edb3e4f5d9b863ae1>

and here is the yml <https://github.com/Prowlarr/Indexers/blob/master/definitions/v3/showrss.yml>

**Comment by Qstick** [comment]
@garfield69 any thoughts here?

**Comment by garfield69** [comment]
Re: 1 title many torrents
Perhaps, rather than reuse the rows design which inherently implies all the data needed is in the iteration of the row block, you could have a group and item block which would make it much clearer that this indexer needs a two-stage iteration
Something like:
```yaml
group:
  selector: div.group
  item: 
    selector: div.torrent
  fields:
    grouptitle:
      selector: div.grouptitle a.name
      attribute: title
      from: group
    title:
      selector: div.itemtitle a.name
      from: item
      filters:
        - name: prepend
          args: "{{ .Result.grouptitle }}"
    details:
      selector: div.grouptitle a.name
      attribute: href
      from: group
    download:
      selector: div.download
      attribute: href
      from: item
```
Or something similar, as fits the actual flow of whatever you C# devs come up with.

If the above is nonsense, then I'm happy to agree with https://github.com/Prowlarr/Indexers/issues/121#issuecomment-1014949535

**Comment by bakerboy448** [comment]
@ilike2burnthing @garfield69 have you guys had a chance to checkout the schemas that Q got setup the other week at all?

One of the few major changes (to support validation) would be for the search results field to contain a `_` if they're not "Cardigann standard"

E.g. `title_french` which is already done, but also for temp results like `random` to `_random`

**Comment by ilike2burnthing** [comment]
Here? https://github.com/Prowlarr/Indexers/pull/206

Looks good. I'll work on updating the indexers, although implementing the schema check will need a little more work.

**Comment by bakerboy448** [comment]
Yup #206 has it! I think there was a fix up or two after that as well

**Comment by ilike2burnthing** [comment]
btmet.yml had `category|append:` changed to `category:` meaning that there are now duplicate `category` keys, which breaks YAML specs (longer discussion here - https://github.com/Jackett/Jackett/issues/11917).

`category|noappend:` is used plenty, so `category|append:` should be fine here.

**Comment by bakerboy448** [comment]
> btmet.yml and funkytorrents.yml had `category|append:` changed to `category:` meaning that there are now duplicate `category` keys, which breaks YAML specs (longer discussion here - [Jackett/Jackett#11917](https://github.com/Jackett/Jackett/issues/11917)).
> 
> `category|noappend:` is used plenty, so `category|append:` should be fine here.

thanks, updated/PR'd
https://github.com/Prowlarr/Indexers/pull/225

**Comment by ilike2burnthing** [comment]
https://github.com/Jackett/Jackett/pull/13376

**Comment by ilike2burnthing** [comment]
Looks like hdai.yml was lost in https://github.com/Prowlarr/Indexers/pull/206. It was deleted from v2 but it doesn't exist anywhere else.

**Comment by bakerboy448** [comment]
> Looks like hdai.yml was lost in #206. It was deleted from v2 but it doesn't exist anywhere else.

thanks! https://github.com/Prowlarr/Indexers/pull/226
